### PR TITLE
[5.7] Runtime Casting for Extended Existentials

### PR DESCRIPTION
**Pull request by CodaFi**
Cherry pick of #59455 and #59507 

Implement the runtime entry points for casting to and from constrained existential types.

rdar://92197049

### CODE REVIEW

**Comment by CodaFi** [comment]
@swift-ci test

File: stdlib/public/runtime/DynamicCast.cpp [hunk 1]
```diff
@@ -2006,12 +2180,14 @@ static tryCastFunctionType *selectCasterForDest(const Metadata *destType) {
...
+    return tryCastToExtendedExistential;
   case MetadataKind::Metatype:
-   return tryCastToMetatype;
- case MetadataKind::ObjCClassWrapper:
+    return tryCastToMetatype;
+  case MetadataKind::ObjCClassWrapper:
     return tryCastToObjectiveCClass;
   case MetadataKind::ExistentialMetatype:
-   return tryCastToExistentialMetatype;
+    return tryCastToExistentialMetatype;
```
Comment (tbkka) [inline stdlib/public/runtime/DynamicCast.cpp:2190]: Thank you for fixing whitespace.

**Review by tbkka — COMMENTED** [review]
I can't comment on the extended existential type theory, but the overall runtime dynamic casting structure looks good.

File: stdlib/public/runtime/DynamicCast.cpp [hunk 2]
```diff
@@ -1695,10 +1695,56 @@ tryCastUnwrappingExistentialSource(
...
+  case ExtendedExistentialTypeShape::SpecialKind::ExplicitLayout: {
+    swift_unreachable("Explicit layout not yet implemented");
+    break;
+  }
+  }
+
+  srcFailureType = srcInnerType;
+  return tryCast(destLocation, destType, srcInnerValue, srcInnerType,
+                 destFailureType, srcFailureType,
+                 takeOnSuccess & (srcInnerValue == srcValue), mayDeferChecks);
```
[the remark this answers is not in the document]
Comment (CodaFi) [reply stdlib/public/runtime/DynamicCast.cpp:1745]: It should be! @rjmccall caught this the first time around. I'll patch this out on main.

**Review by tbkka — APPROVED** [review]
Sorry!  I just realized this is the cherry-pick.  Looks good as-is.  consider my earlier comments suggestions for a follow-up on main.

File: stdlib/public/runtime/DynamicCast.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (CodaFi) [reply stdlib/public/runtime/DynamicCast.cpp:1745]: https://github.com/apple/swift/pull/59515

File: test/Casting/ParameterizedExistentials.swift [hunk 3]
```diff
@@ -0,0 +1,114 @@
+// ParameterizedExistentials.swift - Tests for class constant casts w/ Obj-C
```
[the remark this answers is not in the document]
Comment (CodaFi) [reply test/Casting/ParameterizedExistentials.swift:1]: Good catch.

[the remark this answers is not in the document]
Comment (CodaFi) [reply test/Casting/ParameterizedExistentials.swift:1]: https://github.com/apple/swift/pull/59516

**Pull request merged**